scipy/sparse/bsr.py

```diff
@@ -223,7 +223,10 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False, blocksize=None):
         if dtype is not None:
             self.data = self.data.astype(dtype, copy=False)
 
-        self.check_format(full_check=False)
+        if isinstance(arg1, tuple):
+            self.check_format(full_check=safety_check)
```

Comment:
This feels a little counterintuitive to me. It seems like we actually want to have 3 options for the safety check parameter: never, always, and "auto" (the default).

Reply:
@perimosocordiae Thanks for the feedback. I'm open to other solutions, or enhancements to this solution. When users encounter this bug they only get a segfault, and when sparse matrices are used by other libraries it can make the problem very difficult to debug. Documenting check_format would certainly help, but it does not prevent the undefined behavior from occurring unless users call it (leading to the same performance loss as the proposed solution). The reasons I went with the design of a flag in the constructor are as follows:

- It allows users to turn off the checks at their own risk
- The check only occurs once time, and not during matrix operations after it has been confirmed that the matrix is valid
- Checks only occur when the tuple constructor is used, which is the first opportunity to detect the error, and will only happen when loading a matrix from disk, or when advanced users are constructing a sparse matrix manually

Another solution I thought about is to perform the bounds check in the C++ code, mentioned in bug #8778, and while this would guarantee that no out of bounds reads/writes occur, it would have a significant performance impact.

Reply:
I made the requested change to have a "never", "always", and "auto" safety check mode.